metricbeat/module/haproxy/haproxy.go

```diff
@@ -243,3 +222,86 @@ func (c *Client) GetInfo() (*Info, error) {
 
 	return nil, err
 }
+
+type unixProto struct {
+	Network string
+	Address string
+}
+
+// Run sends a designated command to the haproxy stats socket
+func (p *unixProto) run(cmd string) (*bytes.Buffer, error) {
+	var conn net.Conn
+	response := bytes.NewBuffer(nil)
+
+	conn, err := net.Dial(p.Network, p.Address)
+	if err != nil {
+		return response, err
+	}
+	defer conn.Close()
+
+	_, err = conn.Write([]byte(cmd + "\n"))
+	if err != nil {
+		return response, err
+	}
+
+	_, err = io.Copy(response, conn)
+	if err != nil {
+		return response, err
+	}
+
+	if strings.HasPrefix(response.String(), "Unknown command") {
+		return response, fmt.Errorf("unknown command: %s", cmd)
+	}
+
+	return response, nil
+}
+
+func (p *unixProto) Stat() (*bytes.Buffer, error) {
+	return p.run("show stat")
+}
+
+func (p *unixProto) Info() (*bytes.Buffer, error) {
+	return p.run("show info")
+}
+
+type httpProto struct {
+	URL *url.URL
+}
+
+func (p *httpProto) Stat() (*bytes.Buffer, error) {
+	url := p.URL.String()
+	// Force csv format
+	if !strings.HasSuffix(url, ";csv") {
+		url += ";csv"
+	}
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if p.URL.User != nil {
+		password, _ := p.URL.User.Password()
```

Comment:
I wonder if we should check this error here? Is no password = error?

Reply:
The ignored value is a boolean telling if the password was set, I don't know if there are cases where basic authentication would work with an empty password... In any case I think it can be fine as is, if the credentials are wrong, the request will fail and this error is checked.